Add recursive subcommands

Spectre5 opened the pull request:
Here is an initial take on using recursion for nested subcommands.  I've added tests but I have not well tested it in a "real" application yet.

The docs (help message) as briefly discussed in #89 is not implemented at this point but I do like the idea of adding a `__doc__` or similar method to add help at the "higher level" subcommands.

This implementation also requires the subcommands to be mutable mappings so that the key can be used as the subcommand name.  I'm not sure where the subcommand name would come from if nested lists were used.  Perhaps a "magic" function with `__doc__` in the name could be used for that too, if it is desired to support it.

Closes #89

anntzer commented on test_defopt.py:
```diff
@@ -68,6 +68,81 @@ def sub_with_dash(*, baz=None):
+            return foo
+        # Note that in this case, the commands are all added at the same level
+        # as it isn't clear what the subcommand name would be for the
+        # dictionary entries.
```
In order to move this forward, let's stick to the limited API I propose for now:
- The supported format is a nested dict of names-to-(funcs-or-nested dicts).
- As a special-case, a plain mapping of `{func.__name__: func, ...}` (at any level of nesting) can be replaced by a list of `[func, ...]`.
- In order to allow the "mixed" case, add a `spec` func, basically with the implementation you propose.

The type of `funcs` would then be, I think, the recursive type defined by `T = dict[str, T | Callable] | list[Callable]`.

(I agree that "_1", ... is not particularly helpful.)

Spectre5 replied:
Ok, I will try to implement this soon.  But I want to make sure I understand exact what subset we want to support.  Is this one allowed?

```
defopt.run([sub1, {'sub2': [subsub1, subsub2]}]
```

I think not, since we're disallowing a mapping within a list altogether, right?

anntzer replied:
This should not be allowed.  Basically, each container (list or dict) introduces a new level, and cannot be at the same level as things above it.

Spectre5 commented:
The latest commit disallows mappings within a list.  See the "test_nested_subcommands1", which now must fail, along with "test_nested_subcommands_mixed".

anntzer commented:
Sorry I'm changing things under the rug at the same time.  Can you rebase your PR on top of master instead?

Spectre5 commented:
I did merge in master and I don't see any newer commits.  What else should be added?  I did add a dead-simple example that could potentially be expanded and I added an entry to the changelog.  I guess a new section to the features section of the docs could be added too?

anntzer commented on lib/defopt.py:
```diff
@@ -246,6 +246,38 @@ def run(funcs: Union[Callable, List[Callable], Dict[str, Callable]], *,
+                'A mapping in a list is not allowed, please switch to using '
+                'dictionaries instead.'
+            )
+            # yield from _recurse_functions(func, subparsers)
```
remove?

Spectre5 replied:
Done.

anntzer commented:
Don't worry about the commit history, I'll clean it up when merging the PR.

anntzer commented on lib/defopt.py:
```diff
@@ -246,6 +246,38 @@ def run(funcs: Union[Callable, List[Callable], Dict[str, Callable]], *,
         sys.exit(e)
 
 
+def _recurse_functions(funcs, subparsers):
```
Would it not be easier if you normalized to dicts at the very top, i.e. `if not isinstance(funcs, Mapping): funcs = {func.__name__.replace(...): func for func in funcs}`?  Especially because otherwise the `for func in funcs` iteration can have different meanings depending on the type of funcs.

Spectre5 replied:
Ya, so basically convert any list to a dictionary and then assume we always have a dictionary in the rest of the function.  I can do this.  Gotta run for a bit, but I'll make this change soon!

Spectre5 replied:
Ok, I made this change.  I capture the `AttributeError` that is raised if trying to access __name__ for a list so that I can raise the ValueError instead.  Alternatively, we could leave it as the AttributeError, although then it doesn't give the user any reason on what they did wrong.

anntzer commented:
A blurb in features.rst (e.g. under Subcommands) would be nice, yes.

Spectre5 commented on doc/source/features.rst:
```diff
@@ -79,6 +79,24 @@ Command line usage will use the new names ::
     positional arguments:
       {friendly_func,func2}
 
+Nested Subcommands
```
@anntzer please provide feedback on this -- it is just an initial take.  How do you recommend to document the usage part at the bottom of this section?

anntzer replied:
Something like "the nested subcommands are accessed e.g. by `test.py sub func2`".
Also mention that "subcommands can be nested to an arbitrary level using nested dicts".

anntzer commented on doc/source/features.rst:
```diff
@@ -80,22 +80,16 @@ Command line usage will use the new names ::
       {friendly_func,func2}
 
 Nested Subcommands
-------------------
```
keep underlining as before.

Spectre5 replied:
Ok, I thought you may want it a sub section under the subcommands.  I will promote it.

anntzer commented on doc/source/features.rst:
```diff
@@ -79,6 +79,18 @@ Command line usage will use the new names ::
+Nested Subcommands
+------------------
+
+By using dictionaires, the subcommands can also be nested.
```
dictionaries (typo)

Spectre5 replied:
Fixed.

anntzer commented on doc/source/features.rst:
```diff
@@ -79,6 +79,18 @@ Command line usage will use the new names ::
+    defopt.run({'func1': func1, 'sub': [func2, func3]})
+
+The nested subcommands are accessed e.g. by ``test.py sub func2``.  The
+subcommands can be nested to an arbitrary level by using nested dictionaries.
```
Also link to the runnable example (see `examples/foo.py` links that occur elsewhere in the page).

Spectre5 replied:
Added it.

anntzer commented:
Looks good to me except for one nit.  Also, do you want to squash your commits yourself or should I do it?

Spectre5 commented:
I added the link to the example, added a nested lists example (that should fail), and updated the value error message that is raised when converting `funcs` to a list to read "Use dictionaries (mappings) for nesting; other iterables may only contain functions (callables)."  I think this is slightly more clear to explicit say to use dictionaries for nesting.  The previous message raised here indicated that a dictionary cannot be in a list --- but as the newly added test shows, a list obviously cannot be in a list either!

anntzer commented:
Thanks, merged :)
